client/pwa/src/service-worker.ts

```diff
@@ -122,83 +135,115 @@ export async function messageAllClients(
+  if (updating) {
+    return;
+  }
+
+  const res = await fetch(`${UPDATES_BASE_URL}/update.json`);
+  const update = await res.json();
+
+  if (!update) {
+    console.error(`[refresh] Failed to determine remote version!`, res);
+  }
+
+  await patchContentStatus({
+    remote: {
+      version: update.latest,
+      date: update.date,
+    },
+  });
+}
+
+export async function updateContent(self: ServiceWorkerGlobalScope) {
+  const { local, remote } = await getContentStatus();
+
+  if (!remote || (local && local.version === remote.version)) {
+    return;
+  }
+
   if (updating) {
     return;
   } else {
     updating = true;
   }
+
   try {
-    if (!current) {
+    if (!local) {
+      console.log(`[update] cleaning`);
       await caches.delete(contentCache);
     }
-    await messageAllClients(self, {
-      type: "updateStatus",
-      progress: 0,
-      state: "downloading",
+
+    await patchContentStatus({
+      phase: ContentStatusPhase.download,
     });
 
     const url = new URL(
-      current
-        ? `/packages/${latest}-${current}-update.zip`
-        : `/packages/${latest}-content.zip`,
+      local
+        ? `/packages/${remote.version}-${local.version}-update.zip`
+        : `/packages/${remote.version}-content.zip`,
       UPDATES_BASE_URL
     );
+
     console.log(`[update] downloading: ${url}`);
     const res = await fetch(url.href);
     const buf = await res.arrayBuffer();
 
     console.log(`[update] unpacking: ${url}`);
-    await messageAllClients(self, {
-      type: "updateStatus",
+    await patchContentStatus({
+      phase: ContentStatusPhase.unpack,
       progress: 0,
-      state: "unpacking",
     });
 
     await unpackAndCache(buf, async (progress) => {
-      await messageAllClients(self, {
-        type: "updateStatus",
-        progress,
-        state: "unpacking",
+      await patchContentStatus({
+        phase: ContentStatusPhase.unpack,
+        progress: progress,
       });
     });
 
-    await messageAllClients(self, {
-      type: "updateStatus",
-      progress: 0,
-      state: "init",
-      currentVersion: latest,
-      currentDate: date,
+    await patchContentStatus({
+      phase: ContentStatusPhase.idle,
+      local: remote,
+      progress: null,
     });
 
+    console.log(`[update] synchronizing`);
     await synchronizeDb();
 
     console.log(`[update] done`);
-    updating = false;
   } catch (e) {
-    console.error(e);
+    console.error(`[update] failed`, e);
+  } finally {
     updating = false;
   }
 }
 
+var clearing = false;
+
 async function clearContent(self: ServiceWorkerGlobalScope) {
-  await messageAllClients(self, {
-    type: "updateStatus",
-    progress: 0,
-    state: "clearing",
-  });
-  await caches.delete(contentCache);
-  await messageAllClients(self, {
-    type: "updateStatus",
-    progress: -1,
-    state: "init",
-    currentVersion: null,
-    currentDate: null,
-  });
+  if (clearing) {
+    return;
+  }
+
+  clearing = true;
```

Comment:
Why not. 

`clearing = await getContentStatus().phase === ContentStatusPhase.clear`

Not a change request as such just wondering why not use the db state?

Reply:
Good point!

My first thought was "But what if an error happens?", and that's indeed an issue, because the phase should probably be reset in the finally call, but isn't. 🤔

Reply:
Fixed in 0803bfc72.